department-of-veterans-affairs/caseflow
# Prototype/Add Flipper to Caseflow

~Resolves [Jira Issue Title](https://jira.devops.va.gov/browse/JIRA-12345)~ Prototype only

# Description
This is a prototype branch to test out adding [Flipper](https://www.flippercloud.io/docs/get-started/self-hosted) to Caseflow in order to manage our feature flags.

## Acceptance Criteria

## Testing Plan
1. Create a new feature and add a user as the actor (use the CSS ID as the name) i.e. `BVAISHAW`
2. Check that the feature toggle is enabled for that user via the Rails console
```ruby
# I added BVAISHAW as my sole actor
user = User.find_by_css_id("BVAISHAW")

# My feature name is rails_6_0
Flipper.enabled?(:rails_6_0, user)

# The expected result:
=> true
```

3. Try checking if the feature toggle is enabled for a different user
```ruby
user2 = User.find_by_css_id("JOLLY_POSTMAN")

Flipper.enabled?(:rails_6_0, user2)
=> false
```

4. Register a new group called `vha_users` in the console:
```ruby
Flipper.register(:vha_users) do |actor, context|
  actor.respond_to?(:vha_employee?) && actor.vha_employee?
end
```

5. Add this new group to the config for the feature you've created:
```ruby
Flipper.enable_group(:rails_6_0, :vha_users)
```

6. Check that the UI displays this group as having the feature active:
![group_enabled](https://github.com/department-of-veterans-affairs/caseflow/assets/99351305/80effb71-601f-441f-bea6-de372bff2f36)

7. I can't seem to get the "Add a group" button in the UI to work.. If you can figure it out, please let me know.

8. Grab a VHA user and repeat the check to see if the feature is turned on for them:
```ruby
vha_user = User.find_by_css_id("CAMOUSER")

Flipper.enabled?(:rails_6_0, vha_user)
=> true
```

9. Play around with some of the other configs and let me know if you find anything neat.


# Frontend
## User Facing Changes

![flipper_local_add](https://github.com/department-of-veterans-affairs/caseflow/assets/99351305/56c44fc2-299a-4da0-8340-1aff4009687f)

![flipper_local_all](https://github.com/department-of-veterans-affairs/caseflow/assets/99351305/65cf6cf8-64bc-491b-ae1d-aa8931ceabf8)

# Backend
## Database Changes
*Only for Schema Changes*


# Best practices
## Code Documentation Updates

## Tests
### Test Coverage
Did you include any test coverage for your code? Check below:

### Code Climate
Your code does not add any new code climate offenses? If so why?

## Monitoring, Logging, Auditing, Error, and Exception Handling Checklist

### Monitoring

### Logging

### Auditing

### Error Handling

### Exception Handling
